src/Features/Core/Portable/EmbeddedLanguages/AbstractEmbeddedLanguageCodeFixProvider.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.LanguageServices;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages
+{
+    /// <summary>
+    /// A CodeFixProvider that hooks up the diagnostics produced by <see
+    /// cref="IEmbeddedDiagnosticAnalyzer"/> and to the appropriate <see
+    /// cref="IEmbeddedCodeFixProvider"/>.
+    /// </summary>
+    internal abstract class AbstractEmbeddedLanguageCodeFixProvider : SyntaxEditorBasedCodeFixProvider
+    {
+        private readonly IEmbeddedLanguageProvider _embeddedLanguageProvider;
+        private readonly Dictionary<string, IEmbeddedCodeFixProvider> _diagnosticIdToCodeFixProvider;
+
+        public override ImmutableArray<string> FixableDiagnosticIds { get; }
+
+        protected AbstractEmbeddedLanguageCodeFixProvider(
+            IEmbeddedLanguageProvider embeddedLanguageProvider)
+        {
+            _embeddedLanguageProvider = embeddedLanguageProvider;
+            _diagnosticIdToCodeFixProvider = new Dictionary<string, IEmbeddedCodeFixProvider>();
+
+            // Create a mapping from each IEmbeddedCodeFixProvider.FixableDiagnosticIds back to the
+            // IEmbeddedCodeFixProvider itself.  That way, when we hear about diagnostics, we know
+            // which provider to actually do the fixing.
+            foreach (var language in embeddedLanguageProvider.GetEmbeddedLanguages())
+            {
+                var codeFixProvider = language.CodeFixProvider;
+                if (codeFixProvider != null)
+                {
+                    foreach (var diagnosticId in codeFixProvider.FixableDiagnosticIds)
+                    {
+                        _diagnosticIdToCodeFixProvider[diagnosticId] = codeFixProvider;
```

Comment:
I don't expect that two codefix providers would handle the same diagnostic id, but we sure be resilient (or produce some diagnostic) if there was such a conflict.

Reply:
Sure.  i'll have us throw if that happens.

Reply:
done.